rocks-lib/src/upload/mod.rs

```diff
@@ -0,0 +1,236 @@
+use crate::{config::Config, project::Project};
+use gpgme::{Context, Data, Protocol};
+use reqwest::{
+    multipart::{Form, Part},
+    Client,
+};
+use serde::Deserialize;
+use thiserror::Error;
+
+#[derive(Deserialize, Debug)]
+pub struct VersionCheckResponse {
+    version: String,
+}
+
+#[derive(Error, Debug)]
+pub enum ToolCheckError {
+    #[error(transparent)]
+    Request(#[from] reqwest::Error),
+    #[error("`rocks` is out of date with {0}'s expected tool version! `rocks` is at version {TOOL_VERSION}, server is at {server_version}", server_version = _1.version)]
+    ToolOutdated(String, VersionCheckResponse),
+}
+
+#[derive(Error, Debug)]
+pub enum UserCheckError {
+    #[error(transparent)]
+    Request(#[from] reqwest::Error),
+    #[error("invalid API key provided")]
+    UserNotFound,
+}
+
+#[derive(Error, Debug)]
+#[error("could not check rock status on server: {0}")]
+pub struct RockCheckError(#[from] reqwest::Error);
+
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub enum UploadError {
+    Lua(#[from] mlua::Error),
+    Request(#[from] reqwest::Error),
+    RockCheck(#[from] RockCheckError),
+    #[error("rock already exists on server: {0}")]
+    RockExists(String),
+    #[error("unable to read rockspec: {0}")]
+    RockspecRead(#[from] std::io::Error),
+    Signature(#[from] gpgme::Error),
+    ToolCheck(#[from] ToolCheckError),
+    UserCheck(#[from] UserCheckError),
+}
+
+pub struct ApiKey(String);
+
+#[derive(Error, Debug)]
+#[error("no API key provided! Please set the $ROCKS_API_KEY variable")]
+pub struct ApiKeyUnspecified;
+
+impl ApiKey {
+    /// Retrieves the rocks API key from the `$ROCKS_API_KEY` environment
+    /// variable and seals it in this struct.
+    pub fn new() -> Result<Self, ApiKeyUnspecified> {
+        Ok(Self(
+            env::var("ROCKS_API_KEY").map_err(|_| ApiKeyUnspecified)?,
+        ))
+    }
+
+    /// Creates an API key from a String.
+    ///
+    /// # Safety
+    ///
+    /// This struct is designed to be sealed without a [`Display`](std::fmt::Display) implementation
+    /// so that it can never accidentally be printed.
+    ///
+    /// Ensure that you do not do anything else with the API key string prior to sealing it in this
+    /// struct.
+    pub unsafe fn from(str: String) -> Self {
+        Self(str)
+    }
+
+    /// Retrieves the underlying API key as a [`String`].
+    ///
+    /// # Safety
+    ///
+    /// Strings may accidentally be printed as part of its [`Display`](std::fmt::Display)
+    /// implementation. Ensure that you never pass this variable somewhere it may be displayed.
+    pub unsafe fn get(&self) -> &String {
+        &self.0
+    }
+}
+
+pub async fn upload_from_project(
+    project: &Project,
+    api_key: &ApiKey,
+    config: &Config,
+) -> Result<(), UploadError> {
+    let client = Client::new();
+
+    helpers::ensure_tool_version(&client, config.server()).await?;
+    helpers::ensure_user_exists(&client, api_key, config.server()).await?;
```

Comment:
suggestion: since the API key could be leaked if the URL is not HTTPS, how about we check that it is and return an error if it isn't?

```suggestion

    helpers::ensure_https(config.server())?;
    helpers::ensure_tool_version(&client, config.server()).await?;
    helpers::ensure_user_exists(&client, api_key, config.server()).await?;
```

We could also have `config.server()` perform the check :thinking:

Reply:
Sounds like a plan. Instead of checking the URL string, I'll see if there's a way to force `reqwest` to make HTTPS calls only.